Documentation/update-modules-v3-file-api.md

```diff
@@ -0,0 +1,665 @@
+installed a package.
+
+#### `ArtifactVerifyReboot` state
+
+**[Unimplemented]**
+
+Executes after `ArtifactReboot` has run, if it runs at all.
+
+`ArtifactVerifyReboot` should be used to verify that the reboot has
+been performed correctly, and that it was not rolled back by an external
+component, such as a watch dog or the boot loader. A common use of the script is
+to make sure the correct partition has been booted.
+
+#### `ArtifactCommit` state
+
+Executes after `ArtifactVerifyReboot`, if `ArtifactVerifyReboot` runs at all, or
+else after `ArtifactInstall`. `ArtifactCommit` should be used to make the update
+permanent, in cases where it's still possible to roll back.
+
+#### `Cleanup` state
+
+`Cleanup` executes unconditionally at the end of all the other states,
+regardless of all outcomes. `Cleanup` can be used to cleanup various temporary
+files that might have been used during an update, but should not be used to make
+any system changes. For example, cleaning up an update that has failed,
+returning it to the previous state, should rather be done in the
+`ArtifactRollback` state. `Cleanup` is the only additional state that executes
+if `Download` fails.
+
+### Error states
+
+#### `ArtifactRollback` state
+
+**[Unimplemented]**
+
+`ArtifactRollback` is only considered in some circumstances. Before the
+`Download` state, Mender calls the update module with:
+
+```bash
+./update-module SupportsRollback
+```
+
+where the module can respond with the following responses:
+
+* `No` - Signals that the update module does not support rollback. This is the
+  same as responding with nothing, and hence the default
+* `Yes` - Signals that the update module supports rollback and it should be
+  handled by calling `ArtifactRollback` and `ArtifackRollbackReboot` states
+* `AutomaticDualRootfs` **[Unimplemented]** - Will use the built-in dual rootfs
+  capability of Mender to provide a backup of the currently running system,
+  hence providing a system that can be rolled back to. The module will not be
+  called with the `ArtifactRollback` and `ArtifactRollbackReboot` arguments, but
+  Mender will execute its own internal variants instead. This comes with a few
+  consequences and restrictions:
+  * Only changes to the A/B rootfs partitions can be rolled back
+  * The module cannot stream anything into the inactive partition in the
+    `Download` state, since this partition will be used by Mender to provide the
+    backup
+  * There will be a large delay at the beginning of the update while Mender
+    makes a backup of the current system. Depending on the size of the
+    filesystem, this could be much more time consuming than the update itself
+
+`ArtifactRollback` then executes whenever:
+
+* the `SupportsRollback` call has returned a non-`No` response
+  * For `AutomaticDualRootfs`, only Mender's internal variants are called
+* `ArtifactInstall` has executed successfully
+* `ArtifactReboot`, `ArtifactVerifyReboot` or `ArtifactCommit` fails
+
```

Comment:
I think this documentation requires and update. ArtifactRollback is being called also when ArtifactInstall fails.

Reply:
Thanks, also fixed locally.